File: pattern16_symbol_table_for_monolithic_scope/cpp/Symbol.cpp
#include "Symbol.h"

#include <ostream>
#include <sstream>

std::string Symbol::to_string() const
{
  if (type_)
  {
    std::ostringstream oss;
    oss << '<' << get_name() << ':' << type_->get_name() << '>';
    return oss.str();
  }
  else
    return get_name();
}

std::ostream& operator<<(std::ostream& ostrm, Symbol const& data)
{
  return ostrm << data.to_string();
}

File: pattern16_symbol_table_for_monolithic_scope/java_listener/Symbol.java

public class Symbol
{
  String name;
  Type type;

  public Symbol(String name)
  {
    this.name = name;
  }

  public Symbol(String name, Type type)
  {
    this(name);
    this.type = type;
  }

  public String getName()
  {
    return name;
  }

  public String toString()
  {
    if (type != null)
      return "<" + getName() + ":" + type + ">";
    else
      return getName();
  }
}

